python/orca/tutorial/NCF/pytorch_dataset.py

```diff
@@ -23,69 +23,106 @@
-        super(NCFData, self).__init__()
-        """ Note that the labels are only useful when training, we thus
-            add them in the ng_sample() function.
-        """
-        self.features_ps = features
-        self.num_item = num_item
-        self.train_mat = train_mat
-        self.num_ng = num_ng
-        self.is_sampling = False
-        self.labels = [1 for _ in range(len(features))]
-
-    def ng_sample(self):
-        self.is_sampling = True
-        self.features_ng = []
-        for x in self.features_ps:
-            u = x[0]
-            for t in range(self.num_ng):
-                j = np.random.randint(self.num_item)
-                while (u, j) in self.train_mat:
-                    j = np.random.randint(self.num_item)
-                self.features_ng.append([u, j])
-
-        labels_ps = [1 for _ in range(len(self.features_ps))]
-        labels_ng = [0 for _ in range(len(self.features_ng))]
-
-        self.features_fill = self.features_ps + self.features_ng
-        self.labels_fill = labels_ps + labels_ng
+    def __init__(self, data):
+        self.data = data
 
     def __len__(self):
-        return (self.num_ng + 1) * len(self.labels)
+        return len(self.data)
 
     def __getitem__(self, idx):
-        features = self.features_fill if self.is_sampling \
-            else self.features_ps
-        labels = self.labels_fill if self.is_sampling \
-            else self.labels
-
-        user = features[idx][0]
-        item = features[idx][1]
-        label = float(labels[idx])
-        return user, item, label
-
-
-def load_dataset(dataset_dir):
-    data_X = pd.read_csv(
+        return tuple(self.data[idx])
+
+    # sample negative items for datasets
+    def ng_sampling(self, num_ng, user_num, item_num):
+        # load ratings as a dok matrix
+        features_ps = self.data.values.tolist()
+        train_mat = sp.dok_matrix((user_num, item_num), dtype=np.int64)
+        for x in features_ps:
+            train_mat[x[0], x[1]] = 1
+
+        features_ng = []
+        for x in features_ps:
+            u = x[0]
+            for t in range(num_ng):
+                j = np.random.randint(item_num)
+                while (u, j) in train_mat:
+                    j = np.random.randint(item_num)
+                features_ng.append([u, j])
+        features = features_ps + features_ng
+        labels_ps = [1.0 for _ in range(len(features_ps))]
+        labels_ng = [0.0 for _ in range(len(features_ng))]
+        labels = labels_ps + labels_ng
+        self.data = pd.DataFrame(features, columns=["user", "item"], dtype=np.int64)
+        self.data["label"] = labels
+            
+    def merge_features(self, users, movies, total_cols):
+        self.data = users.merge(self.data, on='user')
+        self.data = self.data.merge(movies, on='item')
+        self.data = self.data.loc[:, total_cols]
+
+
+def load_dataset(dataset_dir, cal_sparse_feats_input_dims=True,
+                 num_ng=4, merge_features=True):
```

Comment:
remove the merge_features options, we will always merge them

Reply:
The option is used for model_creator(), because calculating user_num and item_num in it doesn't need to merge features.

Reply:
I think you can extract a function like `process_user_item` that handles the user/item, which can be used for the model?